shiny/ui/_chat.py

```diff
@@ -573,49 +574,165 @@ async def append_message(
         similar) is specified in model's completion method.
         :::
         """
-        await self._append_message(message, icon=icon)
+        # If we're in a stream, queue the message
+        if self._current_stream_id:
+            self._pending_messages.append((message, False, None))
+            return
+
+        msg = normalize_message(message)
+        msg = await self._transform_message(msg)
+        if msg is None:
+            return
+        self._store_message(msg)
+        await self._send_append_message(
+            message=msg,
+            chunk=False,
+            icon=icon,
+        )
 
-    async def _append_message(
+    @asynccontextmanager
+    async def message_stream_context(self):
+        """
+        Message stream context manager.
+
+        A context manager for appending streaming messages into the chat. This context
+        manager can:
+
+        1. Be used in isolation to append a new streaming message to the chat.
+            * Compared to `.append_message_stream()` this method is more flexible but
+              isn't non-blocking by default (i.e., it doesn't launch an extended task).
+        2. Be nested within itself
+            * Nesting is primarily useful for making checkpoints to `.restore()` back
+              to (see the example below).
+        3. Be used from within a `.append_message_stream()`
+            * Useful for inserting additional content from another context into the
+              stream (e.g., see the note about tool calls below).
+
+        Yields
+        ------
+        :
+            A `MessageStream` class instance, which has a method for `.append()`ing
+            message content chunks to as well as way to `.restore()` the stream back to
+            it's initial state. Note that `.append()` supports the same message content
+            types as `.append_message()`.
+
+        Example
+        -------
+        ```python
+        import asyncio
+
+        from shiny import reactive
+        from shiny.express import ui
+
+        chat = ui.Chat(id="my_chat")
+        chat.ui()
+
+        @reactive.effect
+        async def _():
+            async with chat.message_stream_context() as msg:
+                await msg.append("Starting stream...\n\nProgress:")
+                async with chat.message_stream_context() as progress:
+                    for x in [0, 50, 100]:
+                        await progress.append(f" {x}%")
+                        await asyncio.sleep(1)
+                        await progress.restore()
+                await msg.restore()
```

Comment:
If you're not going to use the term "checkpoint" then I think this should be `msg.clear()`. What do you think?

Reply:
Yea, that is better, thanks 👍